### PR TITLE
feat: add hog harness, a pi distribution

### PR DESCRIPTION
**Pull request by jonathanlab**
## Problem

We need a debuggable, stable, and modifiable harness. 

## Changes

- Added a new package called `harness`, using Pi.dev under the hood. This harness can be launched using either the SDK or the CLI.
- Add a bunch of extensions, some more WIP than others: `subagent`, `web-access`, `posthog-provider`, and `hog-branding`. I've used this community extension for MCP support: https://github.com/nicobailon/pi-mcp-adapter

### CODE REVIEW

File: packages/harness/src/extensions/web-access/web-fetch.ts [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+  if (parts.length < 2) {
+    return { valid: false, reason: "URL must have a public hostname" };
+  }
+
+  return { valid: true, url: parsed };
+}
+
+interface CacheEntry {
+  bytes: number;
+  code: number;
```
[the remark this answers is not in the document]
Comment (jonathanlab) [reply packages/harness/src/extensions/web-access/web-fetch.ts:142]: Fixed in 0fa40448e: `validateUrl` now rejects loopback/private/link-local/CGNAT IPv4 and IPv6 literal hostnames (127.0.0.0/8, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 169.254.0.0/16 incl. cloud metadata, 100.64.0.0/10, 198.18.0.0/15, ::1, fe80::/10, fc00::/7, IPv4-mapped IPv6). Added test coverage including the exact 169.254.169.254 case. Noted in a code comment that this is a literal-address check only — it doesn't protect against DNS rebinding (a public hostname whose DNS record later resolves to a private address), which would need to be enforced at connection time.

File: packages/harness/src/extensions/web-access/web-fetch.ts [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+        return returnMarkdown(cached.content, details);
+      }
+
+      const controller = new AbortController();
+      const timeout = setTimeout(() => controller.abort(), FETCH_TIMEOUT_MS);
+      const combinedSignal = signal
+        ? AbortSignal.any([signal, controller.signal])
+        : controller.signal;
+
```
[the remark this answers is not in the document]
Comment (jonathanlab) [reply packages/harness/src/extensions/web-access/web-fetch.ts:393]: Fixed in 0fa40448e: added readBodyWithLimit(), which streams the response body via response.body.getReader() and enforces MAX_CONTENT_LENGTH as bytes actually arrive, cancelling the reader and throwing once the cap is exceeded — regardless of whether/what the content-length header says. Kept the header-based check as a fast-fail before reading anything when the server is honest about a huge size, but the streaming cap is now the real backstop. Added tests for both an oversized chunked/no-content-length response (rejected) and a small chunked response with no content-length (still works).

File: packages/harness/src/extensions/posthog-provider/oauth.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+      const error = url.searchParams.get("error");
+      res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
+      res.end(SUCCESS_PAGE);
+      cleanup();
+      if (error) {
+        reject(new Error(`PostHog OAuth error: ${error}`));
+      } else if (!code) {
+        reject(new Error("PostHog OAuth callback missing code"));
+      } else if (expectedState && state !== expectedState) {
+        reject(new Error("PostHog OAuth state mismatch"));
```
[the remark this answers is not in the document]
Comment (jonathanlab) [reply packages/harness/src/extensions/posthog-provider/oauth.ts]: Fixed in 0fa40448e, thanks — took your suggestion's approach: the callback handler now determines success/failure first, then serves a distinct failure page ("Authentication failed — please return to your terminal") on error/missing-code/state-mismatch instead of always serving the success page. Added test coverage asserting the served page body for the success path and all three failure paths.

File: packages/harness/src/extensions/subagent/lifecycle.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+ */
+export function writeTranscript(
+  runId: string,
+  markdown: string,
+  maxBytes: number = DEFAULT_MAX_TRANSCRIPT_BYTES,
+): void {
+  const filePath = transcriptPath(runId);
+  fs.mkdirSync(path.dirname(filePath), { recursive: true });
+
+  const byteLength = Buffer.byteLength(markdown, "utf-8");
```
[the remark this answers is not in the document]
Comment (jonathanlab) [reply packages/harness/src/extensions/subagent/lifecycle.ts]: Fixed in 0fa40448e. Style matches our conventions — extracted a shared text-truncate.ts with a single Buffer.from(text, "utf8") encode + at-most-3-byte backoff to the nearest UTF-8 codepoint boundary, used by both lifecycle.ts's writeTranscript and format.ts's truncateForModel (which had the identical O(n) pattern). Added dedicated tests including multi-byte (2-byte é, 4-byte emoji, 3-byte €) boundary cases to confirm no split codepoints/replacement characters.

**Review by k11kirky — APPROVED** [review]
